resources/data_structures_algorithms.md

```diff
@@ -0,0 +1,255 @@
+3. Continue for each element until the array is sorted.
+
+### Selection Sort
+
+Selection Sort improves efficiency by systematically selecting the smallest element from the unsorted segment and moving it to the start. This process repeats with the remaining unsorted section until the array is sorted, with a time complexity of O(n^2).
+
+#### Selection Sort Algorithm Steps:
+
+1. Divide the array into sorted and unsorted segments.
+2. Find and swap the smallest element from the unsorted segment with the first unsorted element.
+3. Repeat until the array is sorted.
+
+### Heap Sort
+
+Heap Sort utilizes a Binary Heap structure, beginning by converting the list into a heap, then repeatedly removing the largest element and maintaining heap properties. It has a time complexity of O(n log n).
+
+#### Heap Sort Algorithm Steps:
+
+1. Convert the array into a heap using the heapify process.
+2. Swap the root of the heap with the last element and remove the last element.
+3. Re-heapify the heap and repeat until one element remains.
+
+## Divide and Conquer Algorithms
+
+### Quick Sort
+
+Quick Sort uses a divide-and-conquer strategy to sort an array by selecting a pivot and partitioning the array around it. It achieves this through recursive partitioning.
+
+#### Quick Sort Algorithm Steps:
+
+1. Select a pivot and partition the array around the pivot.
+2. Recursively apply the same logic to the left and right partitions.
+3. Continue until the array is sorted.
+
+### Merge Sort
+
+Merge Sort divides the array into smaller segments, sorts them, and merges them back into a sorted array. It's a recursive process that ensures stability in sorting.
+
+#### Merge Sort Algorithm Steps:
+
+1. Divide the array into two halves until it cannot be divided further.
+2. Sort each half and merge them back into a single sorted array.
+3. Repeat until the entire array is sorted.
+
+## Graph Algorithms
+
+### Breadth-First Search (BFS)
+
+BFS explores vertices level by level, starting from a specified vertex, making it suitable for finding the shortest path and connected components in graphs.
+
+#### BFS Algorithm Steps:
+
+1. Initialize by placing the starting vertex in a queue and marking it as visited.
+2. Process each vertex from the queue and explore its unvisited neighbors.
+3. Repeat until the queue is empty.
+
+### Depth-First Search (DFS)
+
+DFS explores as deep as possible along each branch before backtracking, used in various graph-related problems.
+
+#### DFS Algorithm Steps:
+
+1. Start at the root and explore along each branch.
+2. Backtrack when no further exploration is possible.
+3. Continue until all vertices are visited.
+
+## Greedy Algorithms in Graphs
+
+### Dijkstra's Algorithm
+
+Dijkstra’s algorithm finds the shortest path from a single vertex to all others by iteratively updating path distances.
+
+### Prim's Algorithm
+
+Prim’s algorithm builds a minimum spanning tree by continuously adding the nearest vertex
```

Comment:
I think some more explanation could be done for each of these topics. For example, maybe explain what a greedy algorithm is?